beacon_node/rest_api/src/beacon_node.rs

```diff
@@ -0,0 +1,83 @@
+use beacon_chain::{BeaconChain, BeaconChainTypes};
+use serde::Serialize;
+use slog::{info, warn};
+use std::sync::Arc;
+use version;
+
+use super::{APIError, APIResult, APIService};
+
+use hyper::{Body, Method, Request, Response, StatusCode};
+use hyper_router::{Route, RouterBuilder};
+
+#[derive(Clone)]
+pub struct BeaconNodeServiceInstance<T: BeaconChainTypes + 'static> {
+    pub marker: std::marker::PhantomData<T>,
+}
+
+/// A string which uniquely identifies the client implementation and its version; similar to [HTTP User-Agent](https://tools.ietf.org/html/rfc7231#section-5.5.3).
+#[derive(Serialize)]
+pub struct Version(String);
+impl ::std::convert::From<String> for Version {
+    fn from(x: String) -> Self {
+        Version(x)
+    }
+}
+
+/// The genesis_time configured for the beacon node, which is the unix time at which the Eth2.0 chain began.
+#[derive(Serialize)]
+pub struct GenesisTime(u64);
+impl ::std::convert::From<u64> for GenesisTime {
+    fn from(x: u64) -> Self {
+        GenesisTime(x)
+    }
+}
+
+impl<T: BeaconChainTypes + 'static> APIService for BeaconNodeServiceInstance<T> {
+    fn add_routes(&mut self, router_builder: RouterBuilder) -> Result<RouterBuilder, hyper::Error> {
+        let router_builder = router_builder
+            .add(Route::get("/version").using(result_to_response!(get_version)))
+            .add(Route::get("/genesis_time").using(result_to_response!(get_genesis_time::<T>)));
+        Ok(router_builder)
+    }
+}
+
+//TODO: Validate request stuff can be turned into a macro.
+fn validate_request(req: &Request<Body>) -> Result<(), APIError> {
```

Comment:
Not a big deal, but it might be handy to move this into `result_to_response!` so we don't end up duplicating it too much.

Another idea might be to create a `handle_get_request!` macro that wraps `result_to_response!` and injects the GET-specific `validate_request(..)`. Then when we have to deal with PUT/PATCH/etc we can have a separate macro.

Reply:
No, I don't think so. I think that each of the Services (in this case the `beacon_node` API service) should have their own validation function. For example, this module only has GET requests, so everything else should fail.
Other modules may have something quite different.

I also intend to integrate any future authentication into this function, which again may be module specific.

Reply:
But don't we need to have separate functions for `Route::put`, etc?

Now I look at this, if we're binding to `Route::get`, is it worthwhile checking that `req.method() == Method::GET`? Perhaps we can just drop the function all together?

Reply:
Yeah, actually you're right. This function is totally redundant in the current code. I had this in there for other purposes in my mind, and from an earlier version of the code where I wasn't using `hyper_router`. I'll pull it out.